plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c

```diff
@@ -0,0 +1,882 @@
+							 sizeof(cmd),
+							 NULL,
+							 0,
+							 FU_GENESYS_GL32XX_DIRECTION_NO_DATA,
+							 error);
+}
+
+static gboolean
+fu_genesys_gl32xx_device_cmd_read_flash(FuGenesysGl32xxDevice *self,
+					guint32 addr,
+					guint8 *data,
+					guint32 datasz,
+					GError **error)
+{
+	guint8 cmd[] = {0xE4, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+	g_return_val_if_fail(data != NULL && datasz != 0, FALSE);
+
+	/* start address */
+	if (!fu_memwrite_uint32_safe(cmd, sizeof(cmd), 2, addr, G_BIG_ENDIAN, error))
+		return FALSE;
+
+	/* block size */
+	if (!fu_memwrite_uint16_safe(cmd, sizeof(cmd), 6, (guint16)datasz, G_BIG_ENDIAN, error))
+		return FALSE;
+
+	return fu_genesys_gl32xx_device_send_scsi_cmd_v3(self,
+							 cmd,
+							 sizeof(cmd),
+							 data,
+							 datasz,
+							 FU_GENESYS_GL32XX_DIRECTION_IN,
+							 error);
+}
+
+static gboolean
+fu_genesys_gl32xx_device_ensure_version(FuGenesysGl32xxDevice *self, GError **error)
+{
+	g_autofree gchar *version = NULL;
+	g_autoptr(GByteArray) buf = NULL;
+
+	buf = fu_genesys_gl32xx_device_cmd_get_version(self, error);
+	if (buf == NULL) {
+		g_prefix_error(error, "failed to read version: ");
+		return FALSE;
+	}
+	if (buf->len < 0x24) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_NOT_SUPPORTED,
+				    "failed to read version");
+		return FALSE;
+	}
+
+	version = g_strdup_printf("%c%c%c%c",
+				  buf->data[0x20],
+				  buf->data[0x21],
+				  buf->data[0x22],
+				  buf->data[0x23]);
+	fu_device_set_version(FU_DEVICE(self), version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_gl32xx_device_check_rom_mode(FuGenesysGl32xxDevice *self,
+					guint8 *cmd,
+					gsize cmdsz,
+					GError **error)
+{
+	const guint8 ext_rom_mode[] = {0x58, 0x52, 0x4F, 0x4D}; /* "XROM" */
+	const guint8 int_rom_mode[] = {0x49, 0x4E, 0x54, 0x2D}; /* "INT-" */
+	guint8 data[4] = {0};
+
+	if (!fu_genesys_gl32xx_device_send_scsi_cmd_v3(self,
+						       cmd,
+						       cmdsz,
+						       data,
+						       sizeof(data),
+						       FU_GENESYS_GL32XX_DIRECTION_IN,
+						       error))
+		return FALSE;
+	if (fu_memcmp_safe(int_rom_mode, sizeof(int_rom_mode), data, sizeof(data), NULL)) {
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		return TRUE;
+	}
+	if (fu_memcmp_safe(ext_rom_mode, sizeof(ext_rom_mode), data, sizeof(data), NULL)) {
+		fu_device_remove_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		return TRUE;
+	}
+
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_FOUND,
+			    "no supported devices detected");
+	return FALSE;
+}
+
+/* safe to call in any mode */
+static gboolean
+fu_genesys_gl32xx_device_ensure_rom_mode(FuGenesysGl32xxDevice *self, GError **error)
+{
+	guint8 cmd[] = {0xE4, 0x01, 0x00, 0x00, 0x04, 0x00};
+	g_autoptr(GError) error_local = NULL;
+
+	g_free(self->chip_name);
+
+	/* check for 3230, 3231, 3232, 3230S, 3231S, 3232S first */
+	cmd[3] = 0xDC;
```

Comment:
maybe better would be:

     	const guint8 cmd_legacy[] = {0xE4, 0x01, 0x00, 0x03, 0x04, 0x00};
     	const guint8 cmd_3224[] = {0xE4, 0x01, 0x00, 0xFC, 0x04, 0x00};